### PR TITLE
Add requires_grad property to Variable

### PR DESCRIPTION
**Pull request by okuta**

### CODE REVIEW

**Comment by niboshi** [comment]
How about adding a test case where variables and arrays are mixed in the function input?

**Comment by okuta** [comment]
Already Chainer tests include that case.

**Comment by niboshi** [comment]
Do you mean `TestFunction.check_call_ndarray()`? OK, I missed it.

**Comment by niboshi** [comment]
LGTM. Could you resolve conflict?

File: chainer/function.py [hunk 1]
```diff
@@ -172,11 +172,16 @@ def __call__(self, *inputs):
...
-                  else variable.Variable(x)
-                  for x in inputs]
+        no_variable = all([not isinstance(x, chainer.Variable)
+                           for x in inputs])
+        if no_variable:
+            in_data = inputs
+        else:
+            inputs = [x if isinstance(x, chainer.Variable)
+                      else chainer.Variable(x)
+                      for x in inputs]
```
Comment (niboshi) [inline chainer/function.py]: Ditto (`variable.Variable`)

File: chainer/function.py [hunk 2]
```diff
@@ -172,11 +172,16 @@ def __call__(self, *inputs):
 
         """
 
-        inputs = [x if isinstance(x, variable.Variable)
-                  else variable.Variable(x)
-                  for x in inputs]
+        no_variable = all([not isinstance(x, chainer.Variable)
+                           for x in inputs])
```
Comment (niboshi) [inline chainer/function.py]: I think `chainer.Variable` should be `variable.Variable`. (rewritten as such in the conflicted commit)

**Comment by okuta** [comment]
I fixed

**Comment by unnonouno** [comment]
Is this feature really OK? Users need to check if a returned value is ndarray or Variable.

**Comment by okuta** [comment]
I can provide helper function.
``` python
def get_data(arg):
    if isinstance(arg, variable.Variable):
        reutrn arg.data
    else:
        return data
```
Is it enough?

**Comment by unnonouno** [comment]
It sounds good. I feel `get_array` is more adequate. And we can provide this function in v1.

**Comment by unnonouno** [comment]
I recently feel it is better to add a flag which represents if a variable is made from ndarrays not from variables.

**Comment by okuta** [comment]
For example `using_config('wrap_ndarray_output`, True)`?

**Comment by unnonouno** [comment]
No. A new attribute for variable.
I think functions must return the same type of values. If it depends on types of arguments, users always need to check return values. I think an additional attribute is enough.

**Comment by okuta** [comment]
I added `requires_grad` property.

This feature is based on  pytorch's similar one.
http://pytorch.org/docs/notes/autograd.html#requires-grad

**Comment by jekbradbury** [comment]
Doesn't this sort of duplicate the behavior we already have with passing ndarrays vs. Variables? Passing an ndarray is equivalent to passing a Variable with requires_grad=False in PyTorch, while passing a Variable is equivalent to passing a Variable with requires_grad=True in PyTorch.

**Comment by okuta** [comment]
I need consistency about return type of `__call__` method. Current implementation returns `variable.Variable` type. And. `variable.Variable` does not have state that means `requires_grad`.

File: chainer/variable.py [hunk 3]
```diff
@@ -264,6 +270,8 @@ class Variable(object):
         update_rule: :class:`~chainer.optimizer.UpdateRule` instance that
             updates this variable as a parameter. This argument is set to
             :attr:`update_rule`.
+        requires_grad (bool): Boolean indicating whether ``grid`` will be set
+            in backward calculation.
```
Comment (niboshi) [inline chainer/variable.py]: Typo: `grid` -> `grad`

**Pull request merged**